### PR TITLE
use protobufs to define ccv state

### PR DESCRIPTION
**Pull request by MSalopek**
Refactors keeping JSON in keeper stores.

Added new messages:
- SlashAcks 
- SlashRequests
- UnbondingOpsIndex
- MaturedUnbondingOps

New message types are used where appropriate, all raw json handling was removed.

Closes https://github.com/cosmos/interchain-security/issues/213

### CODE REVIEW

File: x/ccv/provider/keeper/keeper.go [hunk 1]
```diff
@@ -316,12 +316,12 @@ func (k Keeper) DeleteUnbondingOp(ctx sdk.Context, id uint64) {
 }
 
 // This index allows retreiving UnbondingDelegationEntries by chainID and valsetUpdateID
-func (k Keeper) SetUnbondingOpIndex(ctx sdk.Context, chainID string, valsetUpdateID uint64, IDs []uint64) {
+func (k Keeper) SetUnbondingOpIndex(ctx sdk.Context, chainID string, valsetUpdateID uint64, index ccv.UnbondingOpsIndex) {
```
Comment (jtremback) [inline x/ccv/provider/keeper/keeper.go:319]: Keep the signature `IDs []uint64` and put them into the proto struct in this function.

File: x/ccv/provider/keeper/keeper.go [hunk 2]
```diff
@@ -343,34 +343,32 @@ func (k Keeper) IterateOverUnbondingOpIndex(ctx sdk.Context, chainID string, cb
...
-		if !cb(vscID, ids) {
+		if !cb(vscID, index.GetIds()) {
 			return
 		}
 	}
 }
 
 // This index allows retrieving UnbondingDelegationEntries by chainID and valsetUpdateID
-func (k Keeper) GetUnbondingOpIndex(ctx sdk.Context, chainID string, valsetUpdateID uint64) ([]uint64, bool) {
+func (k Keeper) GetUnbondingOpIndex(ctx sdk.Context, chainID string, valsetUpdateID uint64) (ccv.UnbondingOpsIndex, bool) {
```
Comment (jtremback) [inline x/ccv/provider/keeper/keeper.go:358]: Keep the return type `[]uint64`

File: x/ccv/provider/keeper/keeper.go [hunk 3]
```diff
@@ -433,14 +432,14 @@ func (k Keeper) AppendMaturedUnbondingOps(ctx sdk.Context, ids []uint64) error {
 }
 
 // EmptyMaturedUnbondingOps empties and returns list of matured unbonding operation ids (if it exists)
-func (k Keeper) EmptyMaturedUnbondingOps(ctx sdk.Context) ([]uint64, error) {
```
Comment (jtremback) [inline x/ccv/provider/keeper/keeper.go:436]: Keep `[]uint64` in the signature

File: x/ccv/provider/keeper/keeper.go [hunk 4]
```diff
@@ -398,33 +396,34 @@ func (k Keeper) GetUnbondingOpsFromIndex(ctx sdk.Context, chainID string, valset
 }
 
 // GetMaturedUnbondingOps returns the list of matured unbonding operation ids
-func (k Keeper) GetMaturedUnbondingOps(ctx sdk.Context) (ids []uint64, err error) {
```
Comment (jtremback) [inline x/ccv/provider/keeper/keeper.go:401]: Keep `ids []uint64` in the signature

**Review by jtremback — CHANGES_REQUESTED** [review]
Looks good, but keep the changes inside the getters and setters, so that logic doesn't have to handle these new proto types

File: x/ccv/provider/keeper/keeper.go (hunk 1 above)
Comment (MSalopek) [reply x/ccv/provider/keeper/keeper.go]: `Done.`

File: x/ccv/provider/keeper/keeper.go (hunk 2 above)
Comment (MSalopek) [reply x/ccv/provider/keeper/keeper.go]: `Done.`

File: x/ccv/provider/keeper/keeper.go (hunk 3 above)
Comment (MSalopek) [reply x/ccv/provider/keeper/keeper.go]: `Done.`

File: x/ccv/provider/keeper/keeper.go (hunk 4 above)
Comment (MSalopek) [reply x/ccv/provider/keeper/keeper.go]: `Done.`

**Review by jtremback — APPROVED** [review]

**Review by sainoe — APPROVED** [review]
Good work @MSalopek

**Comment by MSalopek** [comment]
Please wait a bit, there's one more thing to change (SlashRequests). Changes incoming

**Comment by jtremback** [comment]
@MSalopek did you get everything into this that you wanted?

**Comment by MSalopek** [comment]
> @MSalopek did you get everything into this that you wanted?

@jtremback Setting this to draft until proto build changes are merged into main. Will rebase this branch after the changes from #347 are merged.

**Comment by jtremback** [comment]
Getting the "Automated_Tests Expected — Waiting for status to be reported" thing again, but it works locally so i will force merge this.

**Pull request merged**